### PR TITLE
feat: add SQLite support

### PR DESCRIPTION
**Pull request by DanCech**
This PR adds support for using SQLite as a backend database.

## Description

The SQLite storage implementation is largely based on the existing MySQL backend, with some slight differences:
- migration 005 updated to issue separate ALTER TABLE statements for each column, as SQLite does not support multi-column ALTER statements
- date-related SQL updated to use SQLite syntax
- retry mechanism added for writes to avoid SQLITE_BUSY errors (this may not actually be needed with the busy_timeout setting, but is a backup in case the database layer does return SQLITE_BUSY when heavily loaded).
- error handling for SQLite unique constraint errors

It uses the cgo-free https://modernc.org/sqlite driver.

Unless otherwise specified, it will use the [WAL journal mode](https://www.sqlite.org/wal.html) and specify a [busy timeout](https://www.sqlite.org/c3ref/busy_timeout.html) of 500ms

### CODE REVIEW

**Comment by jon-whit** [comment]
@DanCech I changed the base branch this PR is targeting because we're devising a strategy for how to handle community storage contributions without taking on the burden of the storage implementation longer term. We want to promote development, but the maintainer team can't practically support all of them and offer quality implementations if there are a lot of them we're having to maintain.

Consequently, we're thinking of an [openfga/openfga-contrib](https://github.com/openfga/openfga-contrib) repository structure and a plugin based mechanism that will allow the OpenFGA community to run OpenFGA with externally contributed storage implementations. We likely won't merge this into `main` and include it in an official release until we have a more well-defined design for that plugin ecosystem, and in the (hopefully near) future the SQLite implementation you provided herein will be one of the first community sponsored storage implementations in the contrib repo. See https://github.com/openfga/openfga-contrib/blob/main/storage/sqlite/sqlite.go

**Comment by DanCech** [comment]
Thanks for the feedback @jon-whit , it looks like the base change has created quite a few conflicts. Is there anything I can do at this stage to help move things forward?

**Comment by magnusart** [comment]
Hi as a consumer of openfga, if I want to try the SQLite support, what do I need to do at this point? I saw the contrib repo but there were no further instructions as far as I could see?

**Comment by miparnisari** [comment]
Hey @magnusart, we haven't discussed as a team the best path forward on how to add support for more datastores. To unblock you, my recommendation is to fork this repo, and then make sure that you run all the tests like so:

```go
// 1. these tests need an openfga client that is hooked up to a sqllite container; it tests all the important APIs
check.RunAllTests(t, openfgaclient)
listobjects.RunAllTests(t, openfgaclient)
writemodel.RunAllTests(t, openfgaclient)
listusers.RunAllTests(t, openfgaclient)

// 2. this test would need a sqllite container; it runs the integration tests for the storage adapters
openfgateststorage.RunAllTests(t, ds)

// 3. this test would need a sqllite container; it tests all the APIs (yes, again)
openfgatest.RunAllTests(t, ds)
```

This is what we do to test the MySQL and Postgres datastores.

However, it's important to note that the test APIs above are subject to change.

File: assets/migrations/sqlite/001_initialize_schema.sql
Comment (bytefish) [inline assets/migrations/sqlite/001_initialize_schema.sql:4]: As far as I understand SQLite, when defining something like `VARCHAR(255)` everything in parentheses is ignored and it will basically be a `TEXT` column. It’s interesting here to know, that a length of 255 is to be expected, but I would simply make it a `TEXT` and eventually add a `CHECK` constraint to check the length, if one can’t guarantee it from OpenFGA.

Comment (DanCech) [reply assets/migrations/sqlite/001_initialize_schema.sql:4]: Interesting, I have not experimented with CHECK constraints in SQLite. In general I'd tend toward checking length within OpenFGA itself if needed since the behavior of MySQL for example can vary according to the configuration (truncate and warn for non-strict mode, error in strict mode).

**Comment by DanCech** [comment]
Rebased onto current main and updated to account for the storage interface changes, as well as updating to modernc.org/sqlite v1.32.0. I also looked through the diffs between the sqlite driver and tests, and the postgres and mysql implementations, and brought the sqlite support into line as best I could without changing either the mysql or postgres implementations.

I would be happy to open a separate PR to reduce the number of diffs between the drivers so that this process is easier in the future if that would be helpful. Right now there are quite a lot of boilerplate differences and naming inconsistencies to wade through in order to find the real differences.

**Review by miparnisari — COMMENTED** [review]
Couple of other files that need change (either a modification or an addition):
- https://github.com/openfga/openfga/blob/319b81b68c90b4b880261a92b97e37199d63857a/.github/ISSUE_TEMPLATE/report_a_bug.yml#L58
- https://github.com/openfga/openfga/blob/8543ca24d92e4f77fbceaa398a467c591777df1c/.config-schema.json#L174
- https://github.com/openfga/openfga/blob/70440501b780eeea97e9d57042172ffc2f83985e/cmd/migrate/migrate_test.go#L19
- https://github.com/openfga/openfga/blob/a21b3ecf328673ba7a0ebf5eb14b78082f22ccfc/pkg/server/server_test.go#L186
- https://github.com/openfga/openfga/blob/2b8d62a4203c2c46d1613e65c59a59295219b5a6/Makefile#L124-L131
- Tests for ListUsers API https://github.com/openfga/openfga/blob/b40a1e73fa5d845bb1f7f29dca84ac4b255e2e1c/tests/listusers/listusers_test.go#L37-L39
- Tests for Check API https://github.com/openfga/openfga/blob/b40a1e73fa5d845bb1f7f29dca84ac4b255e2e1c/tests/check/check_test.go#L67-L69
- Tests for ListObjects API https://github.com/openfga/openfga/blob/8543ca24d92e4f77fbceaa398a467c591777df1c/tests/listobjects/listobjects_test.go#L22-L24
- Server tests https://github.com/openfga/openfga/blob/a21b3ecf328673ba7a0ebf5eb14b78082f22ccfc/pkg/server/server_test.go#L312-L319
- Metrics tests https://github.com/openfga/openfga/blob/817293c8812835c402dfe1cf7007a18e1d8c04ae/cmd/run/run_test.go#L791-L793
- Create a new benchmark https://github.com/openfga/openfga/blob/a21b3ecf328673ba7a0ebf5eb14b78082f22ccfc/pkg/server/server_test.go#L700-L708

File: pkg/storage/sqlite/sqlite_test.go
Comment (miparnisari) [inline pkg/storage/sqlite/sqlite_test.go:220]: We don't need this test :) @jpadilla can correct me if i'm wrong, but I think this was only applicable to MySQL and Postgres whilst in the middle of the migration from "one record per type" to "one record per model"

File: pkg/storage/sqlite/sqlite_test.go
Comment (miparnisari) [inline pkg/storage/sqlite/sqlite_test.go:224]: I also think we don't need this test. 

Note to self: test if there is a general storage test that asserts that all the `Read*` methods can read tuples with and without conditions correctly

File: pkg/storage/sqlite/sqlite_test.go
Comment (miparnisari) [inline pkg/storage/sqlite/sqlite_test.go:326]: We don't need this test, it was only applicable to MySQL and Postgres https://github.com/openfga/openfga/pull/1217

File: pkg/storage/sqlite/sqlite.go
Comment (miparnisari) [inline pkg/storage/sqlite/sqlite.go:296]: There's an optimization possible in this code, that we didn't have time to bring to the other datastores, but you can bring it here if you like :) https://github.com/openfga/openfga/issues/1873

The gist would be to change:
- the `tuple` table definition
- `Write` method
- All the `Read*` methods involved in reading tuples
- This particular block would be 

```go
if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Relation); ok {
	orConditions = append(orConditions, sq.Eq{"subject_relation": userset.GetRelation()})
	orConditions = append(orConditions, sq.Eq{"subject_type": userset.GetType()})
}
if _, ok := userset.GetRelationOrWildcard().(*openfgav1.RelationReference_Wildcard); ok {
	orConditions = append(orConditions, sq.Eq{"subject_id": "*"})
	orConditions = append(orConditions, sq.Eq{"subject_type": userset.GetType()})
}
```

**Comment by miparnisari** [comment]
Please also open a draft PR to our docs. These pages need update
- https://openfga.dev/docs/getting-started/setup-openfga/configure-openfga#configuring-data-storage
- https://openfga.dev/docs/getting-started/setup-openfga/docker#using-postgres

And finally we also need updates to the Helm chart:
- https://github.com/openfga/helm-charts

File: assets/migrations/sqlite/001_initialize_schema.sql
Comment (DanCech) [reply assets/migrations/sqlite/001_initialize_schema.sql:22]: Interesting, I checked whether there were new migrations but didn't check to see whether existing migrations had changed since I assumed that would never be the case.

File: assets/migrations/sqlite/001_initialize_schema.sql
Comment (DanCech) [reply assets/migrations/sqlite/001_initialize_schema.sql:35]: The default limit for SQLite BLOB is 1GB, so we should be fine there.

https://www.sqlite.org/limits.html

File: assets/migrations/sqlite/002_add_authorization_model_version.sql
Comment (DanCech) [reply assets/migrations/sqlite/002_add_authorization_model_version.sql:1]: I could definitely do that, initially I just wanted to keep things as consistent with the other stores as possible since they are very similar and it makes it easy to compare and check consistency (and the corresponding migrations all have the same number).

File: go.mod
Comment (DanCech) [reply go.mod:53]: Should be good to go I think, it's 3-clause BSD https://gitlab.com/cznic/sqlite/-/blob/master/LICENSE

File: pkg/storage/sqlite/sqlite.go
Comment (DanCech) [reply pkg/storage/sqlite/sqlite.go:296]: Will check it out 👍

File: pkg/storage/sqlite/sqlite_test.go
Comment (DanCech) [reply pkg/storage/sqlite/sqlite_test.go:326]: Cool, I'm all for keeping things clean 😄  Intent here was mostly to demonstrate that this backend has parity with mysql and postgres, and can pass all the same tests.

File: assets/migrations/sqlite/001_initialize_schema.sql
Comment (DanCech) [reply assets/migrations/sqlite/001_initialize_schema.sql:22]: I think I may have misinterpreted your comment earlier. I have kept this backend as close as possible to the existing implementations for MySQL and Postgres (and in fact the only change from the MySQL migrations was to separate 005 into individual ALTER TABLE statements because SQLite doesn't support altering multiple columns in a single transaction.

Now (I think) I understand that you're suggesting jumping straight to the now-current table format vs following the same set of migrations we use for MySQL and PostgreSQL, since the SQLite backend doesn't have to go through that evolution.

File: assets/migrations/sqlite/002_add_authorization_model_version.sql
Comment (DanCech) [reply assets/migrations/sqlite/002_add_authorization_model_version.sql:1]: Switched to a single migration file, luckily goose is ok with us starting at 005 so that the final schema version still matches and all tests still pass,

File: assets/migrations/sqlite/001_initialize_schema.sql
Comment (miparnisari) [reply assets/migrations/sqlite/001_initialize_schema.sql]: Exactly! :)

File: pkg/storage/sqlite/sqlite_test.go
Comment (jpadilla) [reply pkg/storage/sqlite/sqlite_test.go:220]: Yep can drop this for new storage implementations

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (miparnisari) [inline assets/migrations/sqlite/005_initialize_schema.sql:23]: Remove `type_definition` - it's superseded by `serialized_protobuf`

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (miparnisari) [inline assets/migrations/sqlite/005_initialize_schema.sql:26]:
```suggestion
    PRIMARY KEY (store, authorization_model_id)
```

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (miparnisari) [inline assets/migrations/sqlite/005_initialize_schema.sql:24]:
```suggestion
    schema_version VARCHAR(5) NOT NULL DEFAULT '1.1',
```

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (miparnisari) [inline assets/migrations/sqlite/005_initialize_schema.sql:27]:
```suggestion
CREATE TABLE authorization_model (
    store CHAR(26) NOT NULL,
    authorization_model_id CHAR(26) NOT NULL,
    schema_version VARCHAR(5) NOT NULL DEFAULT '1.1',
    serialized_protobuf LONGBLOB NOT NULL,
    PRIMARY KEY (store, authorization_model_id)
);
```

File: go.mod
Comment (miparnisari) [reply go.mod:53]: https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md#approved-licenses-for-allowlist

we're good 👍

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (DanCech) [reply assets/migrations/sqlite/005_initialize_schema.sql:27]: Right now both of these columns are named in the insert query [here](https://github.com/openfga/openfga/blob/23e48b76f6cf24422c6f3517f5d5f9f7fa216ab8/pkg/storage/sqlcommon/sqlcommon.go#L579) so even though they aren't used we cannot drop them from this table without making some supporting changes.

`type_definition` is nullable so we should be able to just remove it from the insert, but `type` is declared not null so the query is explicitly inserting an empty string to keep MySQL and PostgreSQL happy.

One option may be to write migrations to drop those columns from MySQL and PostgreSQL, but there is also the problem of `type` being part of the primary key. It may not be possible to drop it from the primary key if there are entries in older databases that have duplicate values for `store` and `authorization_model_id`.

Comment (miparnisari) [reply assets/migrations/sqlite/005_initialize_schema.sql:27]: > Right now both of these columns are named in the insert query [here](https://github.com/openfga/openfga/blob/23e48b76f6cf24422c6f3517f5d5f9f7fa216ab8/pkg/storage/sqlcommon/sqlcommon.go#L579) so even though they aren't used we cannot drop them from this table without making some supporting changes.

I think it's okay to copy-paste this code into the SQLLite implementation and then modify it according to our needs :)

Comment (DanCech) [reply assets/migrations/sqlite/005_initialize_schema.sql:27]: The only concern I have with doing that is maintainability. Right now the 3 sql backends are *very* similar, so it's relatively easy to roll changes out across all 3 because it's largely a matter of copy+paste+tweak. The more differences creep in, the harder that's going to get long term. I'm happy to go ahead and make a special version of WriteAuthorizationModel for SQLite without the historical baggage if you think that's the best path forward.

File: pkg/storage/sqlite/sqlite_test.go
Comment (DanCech) [reply pkg/storage/sqlite/sqlite_test.go]: Removed :)

**Comment by DanCech** [comment]
Rebased again, I think I've hit all the other files that were identified and I went ahead and removed the `type` and `type_definition` columns for the SQLite backend.

I will look at what's needed to implement #1873 , and will open PRs for docs and helm charts.

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (miparnisari) [inline assets/migrations/sqlite/005_initialize_schema.sql:23]:
```suggestion
    schema_version VARCHAR(5) NOT NULL DEFAULT '1.1',
    serialized_protobuf LONGBLOB NOT NULL,
```

**Comment by DanCech** [comment]
Updated the docs, but I'm not sure what the right path forward is for the helm charts. Running SQLite on k8s doesn't make a huge amount of sense unless you run a single-pod deployment with a PVC to hold the SQLite database files (and ensure that the rollout settings don't allow multiple concurrent pods). It's doable but the noise it would add to the helm chart may not be worth it.

Open to suggestions on how to move forward with that.

**Comment by DanCech** [comment]
@miparnisari the most recent commit includes an implementation of #1873 for SQLite. I based the column names on the code in the linked PRs but can tweak them if you think that the "subject" naming makes more sense.

File: pkg/tuple/tuple.go
Comment (DanCech) [reply pkg/tuple/tuple.go:405]: it's just a couple of small tweaks to the code from #702, so you deserve most of the credit

**Review by jpadilla — APPROVED** [review]
✨ this is great, thank you!

**Comment by DanCech** [comment]
Very cool! I'll rebase again and take care of those merge conflicts

File: CHANGELOG.md
Comment (DanCech) [reply CHANGELOG.md:18]:
```suggestion
* SQLite support (beta) [#1615](https://github.com/openfga/openfga/pull/1615)
```

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (DanCech) [reply assets/migrations/sqlite/005_initialize_schema.sql:1]: goose doesn't actually care, it just wants them in order. The schema version is used internally in a couple of places though so I kept it at 5 to avoid having to make too many changes

File: assets/migrations/sqlite/005_initialize_schema.sql
Comment (miparnisari) [inline assets/migrations/sqlite/005_initialize_schema.sql:19]: If SQLLite supports partial indexes, could we bring these over from Postgres?

```go
CREATE INDEX idx_tuple_partial_user ON tuple (store, object_type, object_id, relation, _user) WHERE user_type = 'user';
CREATE INDEX idx_tuple_partial_userset ON tuple (store, object_type, object_id, relation, _user) WHERE user_type = 'userset';
```

These will make Check API more performant

Comment (DanCech) [reply assets/migrations/sqlite/005_initialize_schema.sql:19]: That should be possible, since SQLite does support partial indexes, I'll check out how that would work together with splitting `_user` up.

**Review by jpadilla — APPROVED** [review]

**Comment by miparnisari** [comment]
We have a flaky test 😢 

image

File: CHANGELOG.md
Comment (miparnisari) [reply CHANGELOG.md:18]: Could you add this back please?

File: README.md
Comment (miparnisari) [reply README.md:20]: same here

**Review by miparnisari — APPROVED** [review]

**Review by jpadilla — APPROVED** [review]

**Pull request merged**
